cpu/kinetis_common/periph/pwm.c

```diff
@@ -24,375 +24,120 @@
-    return (pwm_clk / (1 << prescaler)) / res;
+    /* finally we need to return the actual applied PWM frequency */
+    return (CLOCK_BUSCLOCK >> pre) / res;
 }
 
-uint8_t pwm_channels(pwm_t dev)
+uint8_t pwm_channels(pwm_t pwm)
 {
-    switch (dev) {
-#if PWM_0_EN
-        case PWM_0:
-            return PWM_0_CHANNELS;
-#endif
-#if PWM_1_EN
-        case PWM_1:
-            return PWM_1_CHANNELS;
-#endif
-        default:
-            return 0;
-    }
+    assert(pwm < PWM_NUMOF);
+    return pwm_config[pwm].chan_numof;
 }
 
-void pwm_set(pwm_t dev, uint8_t channel, uint16_t value)
+void pwm_set(pwm_t pwm, uint8_t channel, uint16_t value)
 {
-    FTM_Type *ftm;
-    const uint8_t *ftmchan = NULL;
-
-    switch (dev) {
-#if PWM_0_EN
-
-        case PWM_0:
-            if (channel > PWM_0_CHANNELS) {
-                return;
-            }
-            ftm = PWM_0_DEV;
-            ftmchan = &ftm0chan[0];
-            break;
-#endif
-#if PWM_1_EN
-
-        case PWM_1:
-            if (channel > PWM_1_CHANNELS) {
-                return;
-            }
-            ftm = PWM_1_DEV;
-            ftmchan = &ftm1chan[0];
-            break;
-#endif
-
-        default:
-            return;
-    }
-
-    /* clamp value to maximum possible value */
-    if (value > PWM_MAX_VALUE) {
-        value = PWM_MAX_VALUE;
-    }
-
-    /* cppcheck thinks ftmchan may be NULL here, but the variable is
-     * assigned in all non-returning branches of the switch at the top of
-     * this function. */
-    /* cppcheck-suppress nullPointer */
-    ftm->CONTROLS[ftmchan[channel]].CnV = value;
+    assert((pwm < PWM_NUMOF) && (channel < pwm_config[pwm].chan_numof));
+    ftm(pwm)->CONTROLS[pwm_config[pwm].chan[channel].ftm_chan].CnV = value;
 }
 
-void pwm_start(pwm_t dev)
+void pwm_start(pwm_t pwm)
 {
-    switch (dev) {
-#if PWM_0_EN
-
-        case PWM_0:
-            PWM_0_DEV->SC |= FTM_SC_CLKS(1);
-            break;
-#endif
-#if PWM_1_EN
-
-        case PWM_1:
-            PWM_1_DEV->SC |= FTM_SC_CLKS(1);
-            break;
-#endif
-    }
+    assert(pwm < PWM_NUMOF);
+    ftm(pwm)->SC |= FTM_SC_CLKS(1);
 }
 
-void pwm_stop(pwm_t dev)
+void pwm_stop(pwm_t pwm)
 {
-    switch (dev) {
-#if PWM_0_EN
-
-        case PWM_0:
-            PWM_0_DEV->SC &= ~FTM_SC_CLKS_MASK;
-            break;
-#endif
-#if PWM_1_EN
-
-        case PWM_1:
-            PWM_1_DEV->SC &= ~FTM_SC_CLKS_MASK;
-            break;
-#endif
-    }
+    assert(pwm < PWM_NUMOF);
+    ftm(pwm)->SC &= ~(FTM_SC_CLKS_MASK);
 }
 
-void pwm_poweron(pwm_t dev)
+void pwm_poweron(pwm_t pwm)
 {
-    switch (dev) {
-#if PWM_0_EN
-
-        case PWM_0:
-            PWM_0_CLKEN();
-            break;
-#endif
-#if PWM_1_EN
-
-        case PWM_1:
-            PWM_1_CLKEN();
-            break;
-#endif
+    assert(pwm < PWM_NUMOF);
+    int ftm = pwm_config[pwm].ftm_num;
+    if (ftm < 2) {
+        BITBAND_REG32(SIM->SCGC6, SIM_SCGC6_FTM0_SHIFT + ftm) = 1;
+    }
+    else if (ftm == 2) {
```

Comment:
Add a preprocessor `#if` that checks for the existence of the FTM2 module, it seems like the KW22 doesn't have it which resulted in build failures.